bitcoin/src/bip152.rs

```diff
@@ -116,10 +116,15 @@ impl ShortId {
     }
 
     /// Calculates the short ID with the given (w)txid and using the provided SipHash keys.
-    pub fn with_siphash_keys<T: AsRef<[u8]>>(txid: &T, siphash_keys: (u64, u64)) -> ShortId {
+    ///
+    /// # Parameters
+    ///
+    /// * `txid`: Should be `Txid` or `Wtxid` (eg, use `Txid::as_byte_array()`)
+    /// * `siphash_keys`: The siphash keys (k0, k1).
+    pub fn with_siphash_keys(txid: &[u8], siphash_keys: (u64, u64)) -> ShortId {
```

Comment:
I don't think this helps anything. `AsRef` impl on `Txid`/`Wtxid` should stay but it's true that this method has weak API. I think it'd be more appropriate to have a trait for that that's only implemented for the two expected ids.

```rust
// please suggest a better name
pub trait AsTransactionIdentifier: sealed::Sealed {
    fn as_identifier(&self) -> &[u8; 32];
}
```

Reply:
If we add such a trait we should also add it as a bound on `bitcoin::merkle_node::MerkleNode::Leaf` since the `MerkleNode` trait is currently unsealed and has no bounds on its `Leaf`  type, but it is also intended to only be about wtxid and txid.

Interesting that there are multiple places where we want to be generic over those two types..

Reply:
#2987, a sexy little PR if I don't say so myself. Good ideas crew.